### PR TITLE
[WIP] Add -ConfigurationFile command line parameter to pwsh.exe

### PR DESCRIPTION
**Pull request by PaulHigin**
# PR Summary

This PR adds a new `-ConfigurationFile` command line parameter to pwsh.exe, that takes a path to a .pssc configuration file.
This is a work-in-progress implementation of [RFC](https://github.com/PowerShell/PowerShell-RFC/pull/320).

## PR Context

This new command line parameter will allow PowerShell interactive or server mode sessions to run in a session configured by the provided .pssc file.  

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs [hunk 1]
```diff
@@ -176,6 +176,7 @@ internal static int MaxNameLength()
             "mta",
             "command",
             "configurationname",
+            "configurationfile",
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:179]: ABC-order?

```suggestion
            "configurationfile",
            "configurationname",
```

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs [hunk 2]
```diff
@@ -818,6 +835,18 @@ private void ParseHelper(string[] args)
 
                     _configurationName = args[i];
                 }
+                else if (MatchSwitch(switchKey, "configurationfile", "configurationf"))
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:838]: ABC-order too?

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs [hunk 3]
```diff
@@ -337,6 +338,23 @@ internal Collection<CommandParameter> Args
             }
         }
 
+        internal string? ConfigurationFile
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:341]: ABC-order too?

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs [hunk 4]
```diff
@@ -1646,12 +1653,33 @@ private static bool LoadPSReadline()
         /// Opens and Initializes the Host's sole Runspace.  Processes the startup scripts and runs any command passed on the
         /// command line.
         /// </summary>
-        private void DoCreateRunspace(string initialCommand, bool skipProfiles, bool staMode, string configurationName, Collection<CommandParameter> initialCommandArgs)
+        // Collection<CommandParameter> initialCommandArgs
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs:1656]: Why do we need the comment?

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs (hunk 2 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:838]: It seems it makes no sense to have the short name "configurationf".

File: src/Microsoft.PowerShell.ConsoleHost/resources/ManagedEntranceStrings.resx [hunk 5]
```diff
@@ -288,6 +288,14 @@ All parameters are case-insensitive.</value>
 
     Example: "pwsh -ConfigurationName AdminRoles"
 
+-ConfigurationFile | -configurationf
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/resources/ManagedEntranceStrings.resx:291]: It seems it makes no sense to have the short name "configurationf".

File: src/System.Management.Automation/engine/remoting/fanin/InitialSessionStateProvider.cs [hunk 6]
```diff
@@ -2890,4 +2927,79 @@ internal static string[] TryGetStringArray(object hashObj)
...
+            "VisibleFunctions",
+            "VisibleExternalCommands",
+            "VisibleProviders",
+            "AliasDefinitions",
+            "FunctionDefinitions",
+            "VariableDefinitions",
+            "EnvironmentVariables",
+            "TypesToProcess",
+            "FormatsToProcess",
+            "AssembliesToLoad"
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/remoting/fanin/InitialSessionStateProvider.cs:2963]: Maybe sort the list for readability?

File: src/System.Management.Automation/engine/remoting/fanin/InitialSessionStateProvider.cs [hunk 7]
```diff
@@ -2890,4 +2927,79 @@ internal static string[] TryGetStringArray(object hashObj)
         }
     }
     #endregion
+
+    #region DISCFileValidation
+
+    internal static class DISCFileValidation
+    {
+        // Set of supported configuration options for a PowerShell InitialSessionState.
+        private static readonly HashSet<string> SupportedConfigOptions = new HashSet<string>(StringComparer.InvariantCultureIgnoreCase)
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/remoting/fanin/InitialSessionStateProvider.cs:2936]: Can we use OrdinalIgnoreCase?

File: src/System.Management.Automation/engine/remoting/fanin/InitialSessionStateProvider.cs [hunk 8]
```diff
@@ -2890,4 +2927,79 @@ internal static string[] TryGetStringArray(object hashObj)
...
+            "VariableDefinitions",
+            "EnvironmentVariables",
+            "TypesToProcess",
+            "FormatsToProcess",
+            "AssembliesToLoad"
+        };
+
+        // These are configuration options for WSMan (WinRM) endpoint configurations, that 
+        // appearand in .pssc files, but are not part of PowerShell InitialSessionState.
+        private static readonly HashSet<string> UnsupportedConfigOptions = new HashSet<string>(StringComparer.InvariantCultureIgnoreCase)
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/remoting/fanin/InitialSessionStateProvider.cs:2968]: The same.

File: src/System.Management.Automation/engine/remoting/fanin/InitialSessionStateProvider.cs [hunk 9]
```diff
@@ -2890,4 +2927,79 @@ internal static string[] TryGetStringArray(object hashObj)
...
+        {
+            "RunAsVirtualAccount",
+            "RunAsVirtualAccountGroups",
+            "GroupManagedServiceAccount",
+            "RoleDefinitions",
+            "RequiredGroups",
+            "PowerShellVersion"
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/remoting/fanin/InitialSessionStateProvider.cs:2975]: The same for sorting.

File: src/System.Management.Automation/engine/remoting/fanin/InitialSessionStateProvider.cs [hunk 10]
```diff
@@ -2890,4 +2927,79 @@ internal static string[] TryGetStringArray(object hashObj)
...
+        };
+
+        internal static void ValidateContents(Hashtable configHash)
+        {
+            foreach (var key in configHash.Keys)
+            {
+                if (!(key is string keyName))
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/remoting/fanin/InitialSessionStateProvider.cs:2982]: Does follow work?
```suggestion
                if (key is not string keyName)
```

File: src/System.Management.Automation/engine/remoting/fanin/InitialSessionStateProvider.cs [hunk 11]
```diff
@@ -2890,4 +2927,79 @@ internal static string[] TryGetStringArray(object hashObj)
...
+                    throw new PSInvalidOperationException(RemotingErrorIdStrings.DISCInvalidConfigKeyType);
+                }
+
+                if (!SupportedConfigOptions.Contains(keyName))
+                {
+                    if (UnsupportedConfigOptions.Contains(keyName))
+                    {
+                        throw new PSInvalidOperationException(
+                            StringUtil.Format(RemotingErrorIdStrings.DISCUnsupportedConfigName, keyName));
+                    }
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/remoting/fanin/InitialSessionStateProvider.cs:2993]: If we throw on "unsupported" why do we nest this in `if (!SupportedConfigOptions.Contains(keyName))` condition? Should it has higher priority?

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs [hunk 12]
```diff
@@ -320,6 +321,23 @@ internal Collection<CommandParameter> Args
...
```
Comment (JamesWTruher) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:324]: could you be sure to add telemetry for this new parameter? you may need to rebase.

**Review by JamesWTruher — CHANGES_REQUESTED** [review]

**Comment by SteveL-MSFT** [comment]
Need to re-enable `New-PSSessionConfigurationFile` cmdlet for non-Windows

**Comment by PaulHigin** [comment]
Rather than trying to rebase, I have created a new PR with the changes for this new feature (#17447).

**Pull request closed without merging**